### ISSUE TITLE
How CNN_v2 differs from CNN?

### ISSUE THREAD

**Issue by rqthqwipeh134** [thanh-guong/master-thesis#1]
Hello, I am very interested in your work and would like to ask you a question, compared to the original CNN2, why is your CNN2_v2 designed that way?

**Comment by thanh-guong** [comment]
Hi, thanks for showing your interest and for the question.

CNN2 architecture is explained in [[Convolutional Radio Modulation Recognition
Networks](https://arxiv.org/pdf/1602.04105.pdf)]. CNN2 and CNN_v2 has **the same architecture**, that is shown in the following image.

![Neural Network Architectures-VTCNN2_THESIS drawio(1)](https://user-images.githubusercontent.com/23258665/156737705-c0b119ef-2768-44c7-9929-e5aa03127d25.png)

The only difference between CNN2 and CNN2_v2 is the **implementation**. The CNN2 original implementation is shown in [[this repository](https://github.com/radioML/examples/blob/master/modulation_recognition/RML2016.10a_VTCNN2_example.ipynb)]. That version of the CNN2 implementation **wasn't compatible** (when i tried to run it, I faced a few issues) with the new versions of TensorFlow/Keras, so I did a **refactoring** of the implementation:

- Instead of using _border_mode_ parameter, I used _padding_ in the Convolution2D layers ([docs](https://keras.io/api/layers/convolution_layers/convolution2d/)), and changed the value to "same" instead of "valid"
- Removed the _init_ parameter from Convolution2D layers and Dense layers.

Convolution2D and Dense layers has a default value for _kernel_initializer_ set to "glorot_uniform" ([docs](https://keras.io/api/layers/convolution_layers/convolution2d/)). This refactoring should **only affect Dense** layers, because in the original implementation they have _init_ set to "he_normal".

I hope it helped!

**Comment by gfs32** [comment]
Yes, thanks for such a detailed introduction, the new CNN performs worse, I think it may be related to KERAS's channels_first_or_last

**Comment by rqthqwipeh134** [comment]
Yes, thanks for such a detailed introduction, the new CNN2 performs worse, I think it may be related to KERAS's channels_first_or_last

**Comment by gfs32** [comment]
In addition, the new modulation recognition network has been able to achieve high accuracy, I would like to ask what is the core point of your work? thanks

**Comment by rqthqwipeh134** [comment]
Also, I found that the accuracy would be higher if I normalized the two input data. Recently, new modulation recognition networks have been able to achieve high accuracy. What is the core point of your work? thanks very much

**Comment by thanh-guong** [comment]
The scope of this project is to design a neural network that can be used for **AMR** (Automatic Modulation Recognition).

The objective was to reach **at least 90% accuracy** (distinguing accuracy by SNR level) when trained using the dataset RadioML2016.10a ([ref](https://pubs.gnuradio.org/index.php/grcon/article/view/11)) by RadioML. The accuracy goal **was achieved** with FreeHand v3 and FreeHand v4 neural networks, training them applying some **data augmentation** tecniques (described [here](https://arxiv.org/abs/1912.03026)).

In my thesis work there's a detailed explaination of the results obtained and other stuff. I wrote all in italian language, so it's not readable easily by anyone. If I get authorization I will publish my thesis work here in the repository, and if this interest in my work keeps going, I can consider **translating** the thesis **in english** language.

**Comment by gfs32** [comment]
Oh thanks, don't bother, this FreeHand network and data augmentation technique is amazing, can you recommend some related papers?

**Comment by rqthqwipeh134** [comment]
Oh thanks, no need to bother, this FreeHand network and data augmentation technique works well, can you recommend some related papers?

**Comment by thanh-guong** [comment]
Thanks, I appreciate!

I link here some works that helped me during the thesis work:

| Name                                                                                           | Ref                                                                                                                                     |
|------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------|
| Background, history and an introduction to Automatic Modulation Recognition with Deep Learning | [link](https://www.researchgate.net/publication/332241963_A_Robust_Modulation_Classification_Method_Using_Convolutional_Neural_Networks) |
| Dataset RadioML2016.10a                                                                        | [link](https://pubs.gnuradio.org/index.php/grcon/article/view/11)                                                                        |
| CNN2 neural network                                                                                           | [link](https://arxiv.org/pdf/1602.04105.pdf)                                                                                             |
| RadioML2018.01a, an extended version of the dataset used in my thesis                          | [link](https://ieeexplore.ieee.org/document/8267032)                                                                                     |
| Baseline neural network, relabeling and ensamble learning                                      | [link](https://ieeexplore.ieee.org/document/8935775)                                                                                     |
| Batch normalization                                                                            | [link](https://arxiv.org/abs/1502.03167)                                                                                                 |

**Comment by rqthqwipeh134** [comment]
Thanks a lot!

**Comment by rqthqwipeh134** [comment]
Excuse me, why does (conv(16,4),conv(32,2)) perform better than (conv(256,(1,3)),conv(80,(2,3)))? In this field , I have not seen the convolution kernel size is even. Does it make sense?

**Comment by thanh-guong** [comment]
Hi there, thanks for the question.

This seems to be another question. To avoid confusion between questions, may I ask you to open another issue and moving this question to another issue?

Thanks
